tests/test_gui.py

```diff
@@ -0,0 +1,194 @@
+        [
+            "python",
+            "-m",
+            "tk_toolchain.cmd_line_tools.tk_run_app",
+            # Allows the test for this application to be invoked from
+            # another repository, namely the tk-framework-widget repo,
+            # by specifying that the repo detection should start
+            # at the specified location.
+            "--location",
+            os.path.dirname(__file__),
+            "--context-entity-type",
+            context["type"],
+            "--context-entity-id",
+            str(context["id"]),
+        ]
+    )
+    try:
+        yield
+    finally:
+        # We're done. Grab all the output from the process
+        # and print it so that is there was an error
+        # we'll know about it.
+        stdout, stderr = process.communicate()
+        sys.stdout.write(stdout or "")
+        sys.stderr.write(stderr or "")
+        process.poll()
+        # If returncode is not set, then the process
+        # was hung and we need to kill it
+        if process.returncode is None:
+            process.kill()
+        else:
+            assert process.returncode == 0
+
+
+@pytest.fixture(scope="session")
+def app_dialog(host_application):
+    """
+    Retrieve the application dialog and return the AppDialogAppWrapper.
+    """
+    before = time.time()
+    while before + 30 > time.time():
+        if sgtk.util.is_windows():
+            app_dialog = AppDialogAppWrapper(topwindows)
+        else:
+            app_dialog = AppDialogAppWrapper(topwindows["python"])
+
+        if app_dialog.exists():
+            yield app_dialog
+            app_dialog.close()
+            return
+    else:
+        raise RuntimeError("Timeout waiting for the app dialog to launch.")
+
+
+class AppDialogAppWrapper(object):
+    """
+    Wrapper around the app dialog.
+    """
+
+    def __init__(self, parent):
+        """
+        :param root:
+        """
+        self.root = parent["Shotgun: Publish"].get()
+
+    def exists(self):
+        """
+        ``True`` if the widget was found, ``False`` otherwise.
+        """
+        return self.root.exists()
+
+    def close(self):
+        self.root.buttons["Close"].mouseClick()
+
+
+def test_app(app_dialog):
+    # Make sure items are showing up
+    assert app_dialog.root.captions[
+        "Drag and drop files or folders here"
+    ].exists(), "Drag and drop files or folders here text is missing."
+    assert app_dialog.root.buttons[
+        "Browse files to publish"
+    ].exists(), "Browse files to publish button is missing."
+    assert app_dialog.root.buttons[
+        "Browse sequences to publish"
+    ].exists(), "Browse folders to publish image sequences button is missing."
+
+
+def test_click(app_dialog):
+    # Click on Browse file to publish and then close the file explorer window
+    app_dialog.root.buttons["Browse files to publish"].mouseClick()
+    app_dialog.root.dialogs["Browse files to publish"].waitExist(timeout=30)
```

Comment:
Isn't 30 seconds a bit much?

Reply:
Normally it will never get to the 30 seconds. It will stop as soon as the file browser will show up.
I put 30 seconds to be safe, but I can change it if you want.